### PR TITLE
fix: Push notifications for invite over federation

### PR DESCRIPTION
**Pull request by kate-shine**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by kate-shine** [comment]
Working push notification over federation:
image

**Comment by squahtx** [comment]
There was a previous attempt at fixing this bug at https://github.com/matrix-org/synapse/pull/11410

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -946,6 +946,12 @@ async def on_invite_request(
         )
 
         context = EventContext.for_outlier(self._storage_controllers)
+
+        if event.is_notifiable:
```
Comment (erikjohnston) [inline synapse/handlers/federation.py:950]: I don't think we need this conditional, as it'll always be true (since its an invite)?

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -946,6 +946,12 @@ async def on_invite_request(
...
+            await self._federation_event_handler.bulk_push_rule_evaluator.action_for_event_by_user(
```
Comment (erikjohnston) [inline synapse/handlers/federation.py:951]: We'd prefer if you could add a `self._bulk_push_rule_evaluator = hs.get_bulk_push_rule_evaluator()` to this class than reaching into another's attributes.

File: synapse/handlers/federation_event.py [hunk 3]
```diff
@@ -2173,28 +2174,28 @@ async def persist_events_and_notify(
...
-                        str([ev.event_id for ev in events]),
+                        str([ev.event_id for ev in output_events]),
                     )
                     set_tag(
                         SynapseTags.RESULT_PREFIX + "event_ids.length",
-                        str(len(events)),
+                        str(len(output_events)),
                     )
-                    for event in events:
+                    for event in output_events:
```
Comment (erikjohnston) [inline synapse/handlers/federation_event.py:2198]: Why are these changes necessary?

File: synapse/events/__init__.py [hunk 4]
```diff
@@ -339,12 +339,19 @@ def event_id(self) -> str:
...
 
     def is_state(self) -> bool:
         return self.get_state_key() is not None
 
+    @property
+    def is_notifiable(self) -> bool:
+        return (
+            self.membership == Membership.INVITE
+            or not self.internal_metadata.is_outlier()
+        )
```
Comment (erikjohnston) [inline synapse/events/__init__.py:353]: I'm not a fan of cluttering the `Event` class with more of these "logic" style properties. 

Instead, I think we can move this to `_EventInternalMetadata` as a helper function that does

```python
    def is_notifiable(self) -> bool:
        return not self.is_outlier() or self.is_out_of_band_membership()
```

The `out_band_membership` flag is what we probably want to match on anyway.

**Review by erikjohnston — CHANGES_REQUESTED** [review]
Woo, thanks for this! I think its mostly right, just a few suggestions to make this a bit cleaner.

File: synapse/handlers/federation_event.py (hunk 3 above)
Comment (kate-shine) [reply synapse/handlers/federation_event.py:2198]: Residue of previous attempt for solution I mostly rolled back. I considered changing this back as well, but it improved readability of which events are we currently talking about for me, so I decided to keep it if it doesn't bother anyone. If so, I will change it back.

File: synapse/handlers/federation.py [hunk 5]
```diff
@@ -946,6 +946,12 @@ async def on_invite_request(
...
+                event, context
+            )
+
```
Comment (squahtx) [inline synapse/handlers/federation.py:954]: We'll want to call `remove_push_actions_from_staging` when an exception happens in `persist_events_and_notify`, just like in https://github.com/matrix-org/synapse/blob/0bf180cbb43f6e3d489ddaa5d66226b6daeea027/synapse/handlers/federation_event.py#L2117-L2123

Comment (kate-shine) [reply synapse/handlers/federation.py:954]: Ooof, thanks for noticing! I saw that in the other persist call and thought about putting it here as well, but forgot.

**Comment by squahtx** [comment]
Please feel free to re-request review once this is ready for another look.

File: synapse/events/__init__.py [hunk 6]
```diff
@@ -339,8 +343,8 @@ def event_id(self) -> str:
         raise NotImplementedError()
 
     @property
-    def membership(self) -> str:
-        return self.content["membership"]
+    def membership(self) -> Optional[str]:
+        return self.content.get("membership")
```
Comment (erikjohnston) [inline synapse/events/__init__.py:347]: Why this change? I'm struggling to think through whether this is safe enough to do or not.

**Review by erikjohnston — COMMENTED** [review]
Otherwise, I think this looks good thanks!

File: synapse/events/__init__.py (hunk 6 above)
Comment (kate-shine) [reply synapse/events/__init__.py:347]: Thanks for noticing! It was relevant for the [old version of is_notifiable()](https://github.com/Famedly/synapse/blob/eaf3ab1436c3360f82b9f8b18c14aacfdea8818c/synapse/events/__init__.py#L349), but after requested changes, it's not necessary anymore. It should be absolutely safe, but I will revert it.

**Review by erikjohnston — APPROVED** [review]
Woo, thanks! And thank you for bearing with us during review

**Comment by erikjohnston** [comment]
Err, although the github CI checks haven't run on the latest commit :slightly_frowning_face: Can you merge in develop to see if that kicks it into life?

**Comment by erikjohnston** [comment]
Thanks, that seems to have kicked CI into life! Sorry about that

**Comment by kate-shine** [comment]
> Thanks, that seems to have kicked CI into life! Sorry about that

There was a small conflict there, so I'd have to do it either way.
